Initial data formats for tagging data quality

knopers8 opened the pull request:
Based on the prototypes by @wiechula .

Tagging the interested for a review/discussion: @Barthelemy @chiarazampolli @wiechula (I don't know the usernames of Elena and Francesco).

wiechula committed:
First version for flagging time ranges for analysis

Conflicts:
	DataFormats/CMakeLists.txt

wiechula committed:
Add detector ranges

knopers8 committed:
Reworked FlagReasons and Time-Ranges

ktf reviewed (commented):
I think you are missing the changes to CODEOWNERS for the QualityControl folder. Can you please add them in this or another PR?

knopers8 commented:
indeed, on it

knopers8 committed:
Add codeowners

knopers8 committed:
typo in codeowners

chiarazampolli commented on DataFormats/QualityControl/README.md:
```diff
@@ -0,0 +1,100 @@
+{
+  "name": "CentralBarrelTrackingLimitedAcceptance",
+  "globalRequiredDetectors" : [ "TPC", "ITS" ],
+  "globalIgnoredFlags" : ["Limited acceptance", "asdf"],
```
What does "asdf" mean? It is like "abcd"?

knopers8 replied:
Yes, just a random string.

Barthelemy replied:
(look at a keyboard, middle row)

chiarazampolli replied:
Nice! :-)

chiarazampolli commented on DataFormats/QualityControl/include/DataFormatsQualityControl/FlagReasons.h:
```diff
@@ -0,0 +1,86 @@
+ public:
+  FlagReason& operator=(const FlagReason&) = default;
+  FlagReason(const FlagReason&) = default;
+  bool operator==(const FlagReason& rhs) const;
```
For my curiosity: why "rhs"?

chiarazampolli replied:
Right-Hand-Side :-)

chiarazampolli commented on DataFormats/QualityControl/src/FlagReasons.cxx:
```diff
@@ -0,0 +1,43 @@
+}
+bool FlagReason::operator<(const FlagReason& rhs) const
+{
+  return std::tie(mId, mName, mBad) < std::tie(rhs.mId, rhs.mName, rhs.mBad);
```
Why would we need a less-than and more-than?

knopers8 replied:
For some reason C++ wasn't able to deduce the other operators, even though I would expect that e.g. having `operator<` and `operator==` is enough to derive `operator>`. So I decided it is not worth time investigating for now, but I can have a look again if this bothers you.

chiarazampolli replied:
My question was mainly why do you want to have the "operator<" at all (and the "operator>")?

knopers8 replied:
Ah, it is to be able to sort them in TimeRangeFlagCollection with `std::set`. If we have them sorted, transforming them into Data Tags should be easier. Also, the printed output is easier to understand and with `std::set` merging is for free.

chiarazampolli replied:
Yes, I saw that you want to use std::set to have the merging, I did not know it needed all the operators. 
Your sorting will just be "alphabetical" on the three data members, but it won't mean much, right? Apart from being nice to be seen when printed.
Thanks for explaining.

knopers8 replied:
First we sort by start time, then end time (so chronologically) and after the other data members.

chiarazampolli replied:
Yes, alphabetical included the numbers - maybe one should say it differently. 
Thanks!

chiarazampolli commented on DataFormats/QualityControl/src/TimeRangeFlag.cxx:
```diff
@@ -0,0 +1,60 @@
+  return std::tie(mStart, mEnd, mFlag, mComment, mSource) == std::tie(other.mStart, other.mEnd, other.mFlag, other.mComment, other.mSource);
+}
+
+bool TimeRangeFlag::operator<(const TimeRangeFlag& rhs) const
```
Why would we need also here a less-than and more-than operator?

knopers8 replied:
see above

chiarazampolli commented:
All seems fine to me, I just added a couple of very minor comments. Thanks @knopers8 !!!

shahor02 commented on DataFormats/QualityControl/include/DataFormatsQualityControl/FlagReasons.h:
```diff
@@ -0,0 +1,86 @@
+{
+ private:
+  uint16_t mId;
+  const char* mName;
```
use `std::string` to make it persistent

knopers8 replied:
Ah, I see. ROOT won't store it otherwise?
I used `const char*` so all flags with the same name could point to the same piece of memory. I will see if I can achieve it differently, while allowing for persistent storage.

shahor02 commented on DataFormats/QualityControl/include/DataFormatsQualityControl/TimeRangeFlag.h:
```diff
@@ -0,0 +1,82 @@
+  friend std::ostream& operator<<(std::ostream& output, const TimeRangeFlag& data);
+
+ private:
+  time_type mStart = -1;                          ///< start time of masked range
+  time_type mEnd = -1;                            ///< end time of masked range
```
I would use here something like
````
using RangeInterval = o2::math_utils::Bracket<time_type>;
...
RangeInterval mInterval{};
````
The bracket implements lot of relational methods: https://github.com/AliceO2Group/AliceO2/blob/dev/Common/MathUtils/include/MathUtils/detail/Bracket.h

knopers8 replied:
nice! i will try it out

knopers8 committed:
Use Bracket from MathUtils for the time interval

Barthelemy commented on DataFormats/QualityControl/README.md:
```diff
@@ -0,0 +1,100 @@
+If certain period does not include any TimeRangeFlags with *bad* FlagReasons, then the data quality can be considered as good.
+The [TimeRangeFlagCollection test](test/testTimeRangeFlagCollection.cxx) shows the usage example.
+
+TimeRangeFlagColections are supposed to be created automatically with QC Post-processing Tasks based on Quality Objects created by QC Checks.
```
typo TimeRangeFlagCollections

Barthelemy commented on DataFormats/QualityControl/src/TimeRangeFlagCollection.cxx:
```diff
@@ -0,0 +1,78 @@
+
+#include <iostream>
+
+ClassImp(o2::quality_control::TimeRangeFlagCollection);
```
this is not really useful any more as it was used to generate the THTML for the code documentation. We use doxygen now.

Barthelemy reviewed (dismissed):
Very good. Nice doc. It was educational for me :)

knopers8 committed:
Allow for persistent storage of FlagReasons

Unfortunately, it implies larger memory footprint and no constexpr construction,
 but i don't see other way to achieve it ATM.

knopers8 committed:
Delete unneeded ClassImp and fix typo

knopers8 commented:
All comments are applied.
